File: RadarSDK/RadarSettings.m
//
//  RadarSettings.m
//  RadarSDK
//
//  Copyright © 2019 Radar Labs, Inc. All rights reserved.
//

#import "RadarSettings.h"
#include <Foundation/NSDictionary.h>
#include <Foundation/NSUserDefaults.h>
#include "RadarSdkConfiguration.h"
#include <objc/NSObject.h>

#import "RadarAPIClient.h"
#import "Radar+Internal.h"
#import "RadarLogger.h"
#import "RadarTripOptions.h"
#import "RadarReplayBuffer.h"
#import "RadarLogBuffer.h"
#import "RadarUtils.h"

@implementation RadarSettings

static NSString *const kPublishableKey = @"radar-publishableKey";
static NSString *const kInstallId = @"radar-installId";
static NSString *const kSessionId = @"radar-sessionId";
static NSString *const kId = @"radar-_id";
static NSString *const kUserId = @"radar-userId";
static NSString *const kDescription = @"radar-description";
static NSString *const kProduct = @"radar-product";
static NSString *const kMetadata = @"radar-metadata";
static NSString *const kAnonymous = @"radar-anonymous";
static NSString *const kTracking = @"radar-tracking";
static NSString *const kTrackingOptions = @"radar-trackingOptions";
static NSString *const kPreviousTrackingOptions = @"radar-previousTrackingOptions";
static NSString *const kRemoteTrackingOptions = @"radar-remoteTrackingOptions";
static NSString *const kClientSdkConfiguration = @"radar-clientSdkConfiguration";
static NSString *const kSdkConfiguration = @"radar-sdkConfiguration";
static NSString *const kTripOptions = @"radar-tripOptions";
static NSString *const kLogLevel = @"radar-logLevel";
static NSString *const kBeaconUUIDs = @"radar-beaconUUIDs";
static NSString *const kHost = @"radar-host";
static NSString *const kDefaultHost = @"https://api.radar.io";
static NSString *const kLastTrackedTime = @"radar-lastTrackedTime";
static NSString *const kVerifiedHost = @"radar-verifiedHost";
static NSString *const kDefaultVerifiedHost = @"https://api-verified.radar.io";
static NSString *const kLastAppOpenTime = @"radar-lastAppOpenTime";
static NSString *const kUserDebug = @"radar-userDebug";
static NSString *const kInSurveyMode = @"radar-inSurveyMode";
static NSString *const kXPlatformSDKType = @"radar-xPlatformSDKType";
static NSString *const kXPlatformSDKVersion = @"radar-xPlatformSDKVersion";
static NSString *const kInitializeOptions = @"radar-initializeOptions";
static NSString *const kUserTags = @"radar-userTags";


+ (NSString *)publishableKey {
    return [[NSUserDefaults standardUserDefaults] stringForKey:kPublishableKey];
}

+ (void)setPublishableKey:(NSString *)publishableKey {
    [[NSUserDefaults standardUserDefaults] setObject:publishableKey forKey:kPublishableKey];
}

+ (NSString *)installId {
    NSString *installId = [[NSUserDefaults standardUserDefaults] stringForKey:kInstallId];
    if (!installId) {
        installId = [[NSUUID UUID] UUIDString];
        [[NSUserDefaults standardUserDefaults] setObject:installId forKey:kInstallId];
    }
    return installId;
}

+ (NSString *)sessionId {
    return [NSString stringWithFormat:@"%.f", [[NSUserDefaults standardUserDefaults] doubleForKey:kSessionId]];
}

+ (BOOL)updateSessionId {
    double timestampSeconds = [[NSDate date] timeIntervalSince1970];
    double sessionIdSeconds = [[NSUserDefaults standardUserDefaults] doubleForKey:kSessionId];

    RadarSdkConfiguration *sdkConfiguration = [RadarSettings sdkConfiguration];
    if (sdkConfiguration.extendFlushReplays) {
        [[RadarLogger sharedInstance] logWithLevel:RadarLogLevelInfo type:RadarLogTypeSDKCall message:@"Flushing replays from updateSessionId()"];
        [[RadarReplayBuffer sharedInstance] flushReplaysWithCompletionHandler:nil completionHandler:nil];
    }
    if (timestampSeconds - sessionIdSeconds > 300) {
        [[NSUserDefaults standardUserDefaults] setDouble:timestampSeconds forKey:kSessionId];

        [Radar logOpenedAppConversion];
        
        [[RadarLogger sharedInstance] logWithLevel:RadarLogLevelDebug message:[NSString stringWithFormat:@"New session | sessionId = %@", [RadarSettings sessionId]]];

        return YES;
    }
    return NO;
}

+ (NSString *)_id {
    return [[NSUserDefaults standardUserDefaults] stringForKey:kId];
}

+ (void)setId:(NSString *)_id {
    [[NSUserDefaults standardUserDefaults] setObject:_id forKey:kId];
}

+ (NSString *)userId {
    return [[NSUserDefaults standardUserDefaults] stringForKey:kUserId];
}

+ (void)setUserId:(NSString *)userId {
    NSString *oldUserId = [[NSUserDefaults standardUserDefaults] stringForKey:kUserId];
    if (oldUserId && ![oldUserId isEqualToString:userId]) {
        [RadarSettings setId:nil];
    }
    [[NSUserDefaults standardUserDefaults] setObject:userId forKey:kUserId];
}

+ (NSString *)__description {
    return [[NSUserDefaults standardUserDefaults] stringForKey:kDescription];
}

+ (void)setDescription:(NSString *)description {
    [[NSUserDefaults standardUserDefaults] setObject:description forKey:kDescription];
}

+ (NSString *)product {
    return [[NSUserDefaults standardUserDefaults] stringForKey:kProduct];
}

+ (void)setProduct:(NSString *)product {
    [[NSUserDefaults standardUserDefaults] setObject:product forKey:kProduct];
}

+ (NSDictionary *)metadata {
    return [[NSUserDefaults standardUserDefaults] dictionaryForKey:kMetadata];
}

+ (void)setMetadata:(NSString *)metadata {
    [[NSUserDefaults standardUserDefaults] setObject:metadata forKey:kMetadata];
}

+ (BOOL)anonymousTrackingEnabled {
    return [[NSUserDefaults standardUserDefaults] boolForKey:kAnonymous];
}

+ (void)setAnonymousTrackingEnabled:(BOOL)enabled {
    [[NSUserDefaults standardUserDefaults] setBool:enabled forKey:kAnonymous];
}

+ (BOOL)tracking {
    return [[NSUserDefaults standardUserDefaults] boolForKey:kTracking];
}

+ (void)setTracking:(BOOL)tracking {
    [[NSUserDefaults standardUserDefaults] setBool:tracking forKey:kTracking];
}

+ (RadarTrackingOptions *)trackingOptions {
    NSDictionary *optionsDict = [[NSUserDefaults standardUserDefaults] dictionaryForKey:kTrackingOptions];

    if (optionsDict != nil) {
        return [RadarTrackingOptions trackingOptionsFromDictionary:optionsDict];
    } else {
        // default to efficient preset
        return RadarTrackingOptions.presetEfficient;
    }
}

+ (void)setTrackingOptions:(RadarTrackingOptions *)options {
    NSDictionary *optionsDict = [options dictionaryValue];
    [[NSUserDefaults standardUserDefaults] setObject:optionsDict forKey:kTrackingOptions];
}

+ (void)removeTrackingOptions {
    [[NSUserDefaults standardUserDefaults] removeObjectForKey:kTrackingOptions];
}

+ (RadarTrackingOptions *)previousTrackingOptions {
    NSDictionary *optionsDict = [[NSUserDefaults standardUserDefaults] dictionaryForKey:kPreviousTrackingOptions];

    if (optionsDict != nil) {
        return [RadarTrackingOptions trackingOptionsFromDictionary:optionsDict];
    } else {
        return nil;
    }
}

+ (void)setPreviousTrackingOptions:(RadarTrackingOptions *)options {
    NSDictionary *optionsDict = [options dictionaryValue];
    [[NSUserDefaults standardUserDefaults] setObject:optionsDict forKey:kPreviousTrackingOptions];
}

+ (void)removePreviousTrackingOptions {
    [[NSUserDefaults standardUserDefaults] removeObjectForKey:kPreviousTrackingOptions];
}

+ (RadarTrackingOptions *_Nullable)remoteTrackingOptions {
    NSDictionary *optionsDict = [[NSUserDefaults standardUserDefaults] dictionaryForKey:kRemoteTrackingOptions];

    return optionsDict ? [RadarTrackingOptions trackingOptionsFromDictionary:optionsDict] : nil;
}

+ (void)setRemoteTrackingOptions:(RadarTrackingOptions *_Nonnull)options {
    NSDictionary *optionsDict = [options dictionaryValue];
    [[NSUserDefaults standardUserDefaults] setObject:optionsDict forKey:kRemoteTrackingOptions];
}

+ (void)removeRemoteTrackingOptions {
    [[NSUserDefaults standardUserDefaults] removeObjectForKey:kRemoteTrackingOptions];
}

+ (RadarTripOptions *)tripOptions {
    NSDictionary *optionsDict = [[NSUserDefaults standardUserDefaults] dictionaryForKey:kTripOptions];

    if (optionsDict != nil) {
        return [RadarTripOptions tripOptionsFromDictionary:optionsDict];
    } else {
        return nil;
    }
}

+ (void)setTripOptions:(RadarTripOptions *)options {
    if (options) {
        NSDictionary *optionsDict = [options dictionaryValue];
        [[NSUserDefaults standardUserDefaults] setObject:optionsDict forKey:kTripOptions];
    } else {
        [[NSUserDefaults standardUserDefaults] removeObjectForKey:kTripOptions];
    }
}

+ (NSDictionary *)clientSdkConfiguration {
    NSDictionary *sdkConfigurationDict = [[NSUserDefaults standardUserDefaults] dictionaryForKey:kClientSdkConfiguration];
    if (sdkConfigurationDict == nil) {
        sdkConfigurationDict = [[NSDictionary alloc] init];
    } 
    return sdkConfigurationDict;
}

+ (void) setClientSdkConfiguration:(NSDictionary *)sdkConfiguration {
    if (sdkConfiguration) {
        [[NSUserDefaults standardUserDefaults] setObject:sdkConfiguration forKey:kClientSdkConfiguration];
    } else {
        [[NSUserDefaults standardUserDefaults] removeObjectForKey:kClientSdkConfiguration];
    }
}

+ (void)setSdkConfiguration:(RadarSdkConfiguration *)sdkConfiguration {
    [[RadarLogger sharedInstance] logWithLevel:RadarLogLevelDebug
        message:[NSString stringWithFormat:@"Setting SDK configuration | sdkConfiguration = %@",
                            [RadarUtils dictionaryToJson:[sdkConfiguration dictionaryValue]]]];
    if (sdkConfiguration) {
        [[RadarLogBuffer sharedInstance] setPersistentLogFeatureFlag:sdkConfiguration.useLogPersistence];
        [RadarSettings setLogLevel:sdkConfiguration.logLevel];
        [[NSUserDefaults standardUserDefaults] setObject:[sdkConfiguration dictionaryValue] forKey:kSdkConfiguration];
    } else {
        [[RadarLogBuffer sharedInstance] setPersistentLogFeatureFlag:NO];
        [[NSUserDefaults standardUserDefaults] removeObjectForKey:kSdkConfiguration];
    }
}

+ (RadarSdkConfiguration *_Nullable)sdkConfiguration {
    NSDictionary *sdkConfigurationDict = [[NSUserDefaults standardUserDefaults] dictionaryForKey:kSdkConfiguration];
    return [[RadarSdkConfiguration alloc] initWithDict:sdkConfigurationDict];
}

+ (BOOL)isDebugBuild {
#ifdef DEBUG
    return YES;
#else
    return NO;
#endif
}

+ (RadarLogLevel)logLevel {
    RadarLogLevel defaultLogLevel;
    if ([RadarSettings userDebug]) {
        defaultLogLevel = RadarLogLevelDebug;
    } else if ([self isDebugBuild]) {
        defaultLogLevel = RadarLogLevelInfo;
    } else {
        defaultLogLevel = RadarLogLevelNone;
    }

    if ([[NSUserDefaults standardUserDefaults] objectForKey:kLogLevel] == nil) {
        return defaultLogLevel;
    } 
    return (RadarLogLevel)[[NSUserDefaults standardUserDefaults] integerForKey:kLogLevel];
}

+ (void)setLogLevel:(RadarLogLevel)level {
    [[NSUserDefaults standardUserDefaults] setInteger:(int)level forKey:kLogLevel];
}

+ (NSArray<NSString *> *_Nullable)beaconUUIDs {
    NSArray<NSString *> *beaconUUIDs = [[NSUserDefaults standardUserDefaults] valueForKey:kBeaconUUIDs];
    return beaconUUIDs;
}

+ (void)setBeaconUUIDs:(NSArray<NSString *> *_Nullable)beaconUUIDs {
    [[NSUserDefaults standardUserDefaults] setValue:beaconUUIDs forKey:kBeaconUUIDs];
}

+ (NSString *)host {
    //return @"https://arriving-eagle-magnetic.ngrok-free.app";
    NSString *host = [[NSUserDefaults standardUserDefaults] stringForKey:kHost];
    return host ? host : kDefaultHost;
}

+ (void)updateLastTrackedTime {
    NSDate *timeStamp = [NSDate date];
    [[NSUserDefaults standardUserDefaults] setObject:timeStamp forKey:kLastTrackedTime];
}

+ (NSDate *)lastTrackedTime {
    NSDate *lastTrackedTime = [[NSUserDefaults standardUserDefaults] objectForKey:kLastTrackedTime];
    return lastTrackedTime ? lastTrackedTime : [NSDate dateWithTimeIntervalSince1970:0];
}

+ (NSString *)verifiedHost {
    NSString *verifiedHost = [[NSUserDefaults standardUserDefaults] stringForKey:kVerifiedHost];
    return verifiedHost ? verifiedHost : kDefaultVerifiedHost;
}

+ (BOOL)userDebug {
    NSNumber *userDebug = [[NSUserDefaults standardUserDefaults] objectForKey:kUserDebug];
    return userDebug ? [userDebug boolValue] : NO;
}

+ (void)setUserDebug:(BOOL)userDebug {
    [[NSUserDefaults standardUserDefaults] setBool:userDebug forKey:kUserDebug];
}

+ (void)updateLastAppOpenTime {
    NSDate *timeStamp = [NSDate date];
    [[NSUserDefaults standardUserDefaults] setObject:timeStamp forKey:kLastAppOpenTime];
}

+ (NSDate *)lastAppOpenTime {
    NSDate *lastAppOpenTime = [[NSUserDefaults standardUserDefaults] objectForKey:kLastAppOpenTime];
    return lastAppOpenTime ? lastAppOpenTime : [NSDate dateWithTimeIntervalSince1970:0];
}

+ (BOOL)useRadarModifiedBeacon {
    return [[self sdkConfiguration] useRadarModifiedBeacon];
}

+ (BOOL)xPlatform {
    return [[NSUserDefaults standardUserDefaults] stringForKey:kXPlatformSDKType] != nil &&
    [[NSUserDefaults standardUserDefaults] stringForKey:kXPlatformSDKVersion];
}
+ (NSString *)xPlatformSDKType {
    return [[NSUserDefaults standardUserDefaults] stringForKey:kXPlatformSDKType];
}
+ (NSString *)xPlatformSDKVersion {
    return [[NSUserDefaults standardUserDefaults] stringForKey:kXPlatformSDKVersion];
}

+ (BOOL)useOpenedAppConversion {
    if (![self sdkConfiguration]) {
        return YES;
    }
    return [[self sdkConfiguration] useOpenedAppConversion];
}

+ (void)setInitializeOptions:(RadarInitializeOptions *)options {
    [[NSUserDefaults standardUserDefaults] setObject:[options dictionaryValue] forKey:kInitializeOptions];
}

+ (RadarInitializeOptions *)initializeOptions {
    NSDictionary *dict = [[NSUserDefaults standardUserDefaults] dictionaryForKey:kInitializeOptions];
    if (!dict) {
        return nil;
    }
    return [[RadarInitializeOptions alloc] initWithDict:dict];
}

+ (BOOL)isInSurveyMode {
    return [[NSUserDefaults standardUserDefaults] boolForKey:kInSurveyMode];
}

+ (void)setInSurveyMode:(BOOL)inSurveyMode {
    [[NSUserDefaults standardUserDefaults] setBool:inSurveyMode forKey:kInSurveyMode];
}


+ (NSArray<NSString *> *_Nullable)tags {
    return [[NSUserDefaults standardUserDefaults] arrayForKey:kUserTags];
}

+ (void)setTags:(NSArray<NSString *> *_Nullable)tags {
    if (tags) {
        [[NSUserDefaults standardUserDefaults] setObject:tags forKey:kUserTags];
    } else {
        [[NSUserDefaults standardUserDefaults] removeObjectForKey:kUserTags];
    }
}

+ (void)addTags:(NSArray<NSString *> *_Nonnull)tags {
    NSMutableArray<NSString *> *existingTags = [[self tags] mutableCopy];
    if (!existingTags) {
        existingTags = [NSMutableArray new];
    }
    
    NSSet<NSString *> *existingTagsSet = [NSSet setWithArray:existingTags];
    for (NSString *tag in tags) {
        if (![existingTagsSet containsObject:tag]) {
            [existingTags addObject:tag];
        }
    }
    
    [[NSUserDefaults standardUserDefaults] setObject:existingTags forKey:kUserTags];
}

+ (void)removeTags:(NSArray<NSString *> *_Nonnull)tags {
    NSMutableArray<NSString *> *existingTags = [[self tags] mutableCopy];
    if (!existingTags) {
        return;
    }
    
    [existingTags removeObjectsInArray:tags];
    
    if (existingTags.count > 0) {
        [[NSUserDefaults standardUserDefaults] setObject:existingTags forKey:kUserTags];
    } else {
        [[NSUserDefaults standardUserDefaults] removeObjectForKey:kUserTags];
    }
}

@end

File: RadarSDK/RadarSettings.swift
//
//  RadarSettings.swift
//  RadarSDK
//
//  Copyright © 2025 Radar Labs, Inc. All rights reserved.
//

import Foundation

@objc(RadarSettings) @objcMembers
internal class RadarSettings: NSObject {  // swiftlint:disable:this type_body_length

    static let DefaultHost = "https://api.radar.io"
    static let DefaultVerifiedHost = "https://api-verified.radar.io"
    static let DefaultVerifiedHostSecondary = "https://api-verified.radar.com"

    public static var defaultVerifiedHostSecondary: String { DefaultVerifiedHostSecondary }

    public static func setAppGroup(_ appGroup: String?) {
        // this call needs to by synchronised to prevent race conditions in checking / updating the app group
        let updateAppGroup = {
            // no change in app group
            if RadarUserDefaults.string(forKey: .appGroup) == appGroup {
                return
            }

            let prevUserDefaults = RadarUserDefaults.userDefaults
            // if new user defaults cannot be created, default to .standard, UserDefaults(suiteName: nil) also references the standard user default
            let newUserDefaults = UserDefaults(suiteName: appGroup) ?? .standard

            // if newUserDefaults[AppGroup] is already equal to the appGroup, we're already cloned the UserDefaults,
            // so we just start using the new user defaults. This is for initializing with an app group.
            if newUserDefaults.string(forKey: RadarUserDefaults.Key.appGroup.rawValue) == appGroup {
                UserDefaults.standard.set(appGroup, forKey: RadarUserDefaults.Key.appGroup.rawValue)
                RadarUserDefaults.userDefaults = newUserDefaults
                return
            }

            RadarUserDefaults.clone(from: prevUserDefaults, to: newUserDefaults)

            // set AppGroup to nil, so next time we switch from another app group to the current, it'll detect it's new and clone
            RadarUserDefaults.set(nil, forKey: .appGroup)
            RadarUserDefaults.userDefaults = newUserDefaults

            // set AppGroup in default keys so we can initialize to the correct app group
            UserDefaults.standard.set(appGroup, forKey: RadarUserDefaults.Key.appGroup.rawValue)
            // and set appGroup in new user defaults to signify initialized
            RadarUserDefaults.set(appGroup, forKey: .appGroup)
        }

        if Thread.isMainThread {
            updateAppGroup()
        } else {
            DispatchQueue.main.sync {
                updateAppGroup()
            }
        }
    }

    public static func getAppGroup() -> String? {
        return RadarUserDefaults.string(forKey: .appGroup)
    }

    public static var publishableKey: String? {
        get { return RadarUserDefaults.string(forKey: .publishableKey) }
        set { RadarUserDefaults.set(newValue, forKey: .publishableKey) }
    }

    public static var installId: String {
        if let uuid = RadarUserDefaults.string(forKey: .installId) {
            return uuid
        } else {
            let uuid = UUID().uuidString
            RadarUserDefaults.set(uuid, forKey: .installId)
            return uuid
        }
    }

    public static var sessionId: String {
        String(format: "%.f", RadarUserDefaults.double(forKey: .sessionId))
    }

    public static func updateSessionId() -> Bool {
        let timestampSeconds: Double = Date().timeIntervalSince1970
        var sessionIdSeconds: Double = RadarUserDefaults.double(forKey: .sessionId)

        let sdkConfiguration = RadarSettings.sdkConfiguration
        if sdkConfiguration?.extendFlushReplays ?? false {
            RadarLogger.shared.info("Flushing replays from updateSessionId()", type: .sdkCall)
            // TODO: call swift RadarReplayBuffer when implemented
            RadarSwift.bridge?.flushReplays()
        }

        if timestampSeconds - sessionIdSeconds > 300 {
            sessionIdSeconds = timestampSeconds
            RadarUserDefaults.set(sessionIdSeconds, forKey: .sessionId)
            // TODO: fix this call when it can be called from swift cleanly
            RadarSwift.bridge?.logOpenedAppConversion()
            RadarLogger.shared.debug(String(format: "New session | sessionId = %@", RadarSettings.sessionId))
            return true
        }
        return false
    }

    public static var id: String? {
        @objc(_id)
        get { return RadarUserDefaults.string(forKey: .id) }
        set { RadarUserDefaults.set(newValue, forKey: .id) }
    }

    public static var userId: String? {
        get { return RadarUserDefaults.string(forKey: .userId) }
        set {
            let oldUserId = RadarUserDefaults.string(forKey: .userId)
            if oldUserId != nil && oldUserId != newValue {
                RadarSettings.id = nil
            }
            RadarUserDefaults.set(newValue, forKey: .userId)
        }
    }

    public static var userLanguage: String? {
        get { return RadarUserDefaults.string(forKey: .userLanguage) }
        set { RadarUserDefaults.set(newValue, forKey: .userLanguage) }
    }

    public static var description: String? {
        @objc(__description)
        get { return RadarUserDefaults.string(forKey: .description) }
        set { RadarUserDefaults.set(newValue, forKey: .description) }
    }

    public static var product: String? {
        get { return RadarUserDefaults.string(forKey: .product) }
        set { RadarUserDefaults.set(newValue, forKey: .product) }
    }

    public static var metadata: [String: Any]? {
        get { return RadarUserDefaults.dictionary(forKey: .metadata) }
        set { RadarUserDefaults.set(newValue, forKey: .metadata) }
    }

    public static var anonymousTrackingEnabled: Bool {
        get { return RadarUserDefaults.bool(forKey: .anonymous) }
        set { RadarUserDefaults.set(newValue, forKey: .anonymous) }
    }

    public static var tracking: Bool {
        get { return RadarUserDefaults.bool(forKey: .tracking) }
        set { RadarUserDefaults.set(newValue, forKey: .tracking) }
    }

    public static var trackingOptions: RadarTrackingOptions! {
        get {
            if let optionsDict = RadarUserDefaults.dictionary(forKey: .trackingOptions) {
                return RadarTrackingOptions(from: optionsDict) ?? .presetEfficient
            }
            return .presetEfficient
        }
        set {
            RadarUserDefaults.set(newValue?.dictionaryValue(), forKey: .trackingOptions)
        }
    }

    public static var previousTrackingOptions: RadarTrackingOptions? {
        get {
            if let options = RadarUserDefaults.dictionary(forKey: .previousTrackingOptions) {
                return RadarTrackingOptions(from: options)
            }
            return nil
        }
        set {
            RadarUserDefaults.set(newValue?.dictionaryValue(), forKey: .previousTrackingOptions)
        }
    }

    public static var remoteTrackingOptions: RadarTrackingOptions? {
        get {
            if let options = RadarUserDefaults.dictionary(forKey: .remoteTrackingOptions) {
                return RadarTrackingOptions(from: options)
            }
            return nil
        }
        set { RadarUserDefaults.set(newValue?.dictionaryValue(), forKey: .remoteTrackingOptions) }
    }

    public static var tripOptions: RadarTripOptions? {
        get {
            if let options = RadarUserDefaults.dictionary(forKey: .tripOptions) {
                return RadarTripOptions(from: options)
            }
            return nil
        }
        set { RadarUserDefaults.set(newValue?.dictionaryValue(), forKey: .tripOptions) }
    }

    public static var trip: RadarTrip? {
        get {
            if let dict = RadarUserDefaults.dictionary(forKey: .trip) {
                return RadarTrip(object: dict)
            }
            return nil
        }
        set { RadarUserDefaults.set(newValue?.dictionaryValue(), forKey: .trip) }
    }

    public static var clientSdkConfiguration: [String: Any] {
        get { return RadarUserDefaults.dictionary(forKey: .clientSdkConfiguration) ?? [:] }
        set { RadarUserDefaults.set(newValue, forKey: .clientSdkConfiguration) }
    }

    public static var sdkConfiguration: RadarSdkConfiguration? {
        get {
            let options = RadarUserDefaults.dictionary(forKey: .sdkConfiguration)
            return RadarSdkConfiguration(dict: options)
        }
        set {
            RadarUserDefaults.set(newValue?.dictionaryValue(), forKey: .sdkConfiguration)

            if let newValue {
                logLevel = newValue.logLevel
            }
        }
    }

    public static var logLevel: RadarLogLevel {
        get {
            if RadarUserDefaults.object(forKey: .logLevel) == nil {
                if userDebug {
                    return .debug
                } else {
                    #if DEBUG
                        return .debug
                    #else
                        return .none
                    #endif
                }
            }
            return RadarLogLevel(rawValue: RadarUserDefaults.integer(forKey: .logLevel)) ?? .none
        }
        set { RadarUserDefaults.set(newValue.rawValue, forKey: .logLevel) }
    }

    public static var beaconUUIDs: [String]? {
        get { return RadarUserDefaults.object(forKey: .beaconUUIDs) as? [String] }
        set { RadarUserDefaults.set(newValue, forKey: .beaconUUIDs) }
    }

    public static var host: String {
        return RadarUserDefaults.string(forKey: .host) ?? DefaultHost
    }

    public static var lastTrackedTime: Date {
        let lastTrackedTime: Date? = RadarUserDefaults.object(forKey: .lastTrackedTime) as? Date
        return lastTrackedTime ?? Date(timeIntervalSince1970: 0)
    }

    public static var verifiedHost: String {
        RadarUserDefaults.string(forKey: .verifiedHost) ?? DefaultVerifiedHost
    }

    public static var userDebug: Bool {
        get {
            return RadarUserDefaults.bool(forKey: .userDebug)
        }
        set {
            RadarUserDefaults.set(newValue, forKey: .userDebug)
        }
    }

    public static var lastAppOpenTime: Date {
        let lastAppOpenTime: Date? = RadarUserDefaults.object(forKey: .lastAppOpenTime) as? Date
        return lastAppOpenTime ?? Date(timeIntervalSince1970: 0)
    }

    public static var useRadarModifiedBeacon: Bool {
        sdkConfiguration?.useRadarModifiedBeacon ?? false
    }

    public static var xPlatform: Bool {
        xPlatformSDKType != nil && xPlatformSDKVersion != nil
    }

    public static var xPlatformSDKType: String? {
        RadarUserDefaults.string(forKey: .xPlatformSDKType)
    }

    public static var xPlatformSDKVersion: String? {
        RadarUserDefaults.string(forKey: .xPlatformSDKVersion)
    }

    public static var useOpenedAppConversion: Bool {
        sdkConfiguration?.useOpenedAppConversion ?? true
    }

    public static var initializeOptions: RadarInitializeOptions? {
        get {
            if let options = RadarUserDefaults.dictionary(forKey: .initializeOptions) {
                return RadarInitializeOptions(dict: options)
            }
            return nil
        }
        set { RadarUserDefaults.set(newValue?.dictionaryValue(), forKey: .initializeOptions) }
    }

    public static var inSurveyMode: Bool {
        get { return RadarUserDefaults.bool(forKey: .inSurveyMode) }
        set { RadarUserDefaults.set(newValue, forKey: .inSurveyMode) }
    }

    public static var pushNotificationToken: String? {
        get { return RadarUserDefaults.string(forKey: .pushNotificationToken) }
        set { RadarUserDefaults.set(newValue, forKey: .pushNotificationToken) }
    }

    public static var locationExtensionToken: String? {
        get { return RadarUserDefaults.string(forKey: .locationExtensionToken) }
        set { RadarUserDefaults.set(newValue, forKey: .locationExtensionToken) }
    }

    public static func updateLastTrackedTime() {
        RadarUserDefaults.set(Date(), forKey: .lastTrackedTime)
    }

    public static func updateLastAppOpenTime() {
        RadarUserDefaults.set(Date(), forKey: .lastAppOpenTime)
    }

    public static var tags: [String] {
        get { return RadarUserDefaults.array(forKey: .userTags) as? [String] ?? [] }
        set { RadarUserDefaults.set(newValue, forKey: .userTags) }
    }

    public static func addTags(_ tags: [String]) {
        var existingTags: [String] = self.tags
        let existingTagsSet: Set<String> = Set(self.tags)
        for tag in tags {
            if !existingTagsSet.contains(tag) {
                existingTags.append(tag)
            }
        }
        self.tags = existingTags
    }

    public static func removeTags(_ tags: [String]) {
        self.tags = self.tags.filter { !tags.contains($0) }
    }
}
